@navikt/core/react/src/form/file-upload/FileUpload.tsx

```diff
@@ -0,0 +1,151 @@
+import { partitionFiles } from "./partition-files";
+import ButtonVariant from "./ButtonVariant";
+import ZoneVariant from "./ZoneVariant";
+
+export interface OnUploadProps {
+  allFiles: File[],
+  acceptedFiles: File[],
+  rejectedFiles: File[]
+}
+
+export interface FileUploadProps {
+  /**
+   * Changes styling when changed.
+   * The "zone" variant takes up the full
+   * width of its parent, while the "button"
+   * variant takes up width based on its content.
+   * Both styles support opening a file viewer
+   * and drag-and-drop.
+   */
+  variant: "zone" | "button";
+  /**
+   * ID of the input element. Required to properly
+   * connect input element to potential error message.
+   */
+  inputId: string;
+  /**
+   * Text shown to the user.
+   */
+  label: string;
+  /**
+   * Adds a description to extend labling of FileUpload
+   */
+  description?: string;
+  /**
+   * Class name passed to the outermost <div> element.
+   */
+  className?: string;
+  /**
+   * Error message.
+   */
+  error?: string;
+  /**
+   * Indicates if it is possible
+   * to upload multiple files at once.
+   */
+  multiple?: boolean;
+  /**
+   * Indicates which MIME types are
+   * selectable in the file browser.
+   */
+  accept?: string;
+  /**
+   * Sets a ref on the outermost div.
+   */
+  ref?: React.Ref<HTMLDivElement>;
+  /**
+   * Callback triggered on file upload
+   */
+  onUpload({ allFiles, acceptedFiles, rejectedFiles }: OnUploadProps): void;
+  /**
+   * Custom validator that is used to decide
+   * if a file is accepted or rejected.
+   */
+  validator?(file: File): boolean;
+}
+
+/**
+ * A component for uploading files
+ * @see [📝 Documentation](https://aksel.nav.no/komponenter/core/FileUpload)
+ * @see 🏷️ {@link FileUploadProps}
+ * @example
+ * ```jsx
+ * <FileUpload
+ *   onUpload={onUpload}
+ *   variant="button"
+ *   label="Last opp filer her"
+ *   inputId="fileupload-input"
+ * />
+ * ```
+ */
+export const FileUpload = forwardRef<HTMLDivElement, FileUploadProps>(
+  (
+    {
+      onUpload,
+      error,
+      inputId,
+      label,
+      description,
+      className,
+      variant = "zone",
+      multiple = true,
+      accept,
+      validator
+    },
+    ref
+  ) => {
+    const [isDraggingOver, setIsDraggingOver] = useState<boolean>(false)
+
+    const onDragEnter = () => setIsDraggingOver(true)
+    const onDragEnd = () => setIsDraggingOver(false)
+
+    const handleUpload = (event: ChangeEvent<HTMLInputElement>) => {
+      const fileList = event.target.files
+      if (!fileList) {
+        return
+      }
+      const files = Array.from(fileList)
+      const { acceptedFiles, rejectedFiles } = partitionFiles(files, accept, validator)
+
+      onUpload({ allFiles: files, acceptedFiles, rejectedFiles })
+
+      // Resets the value to make it is possible to upload the same file several consecutive times
+      event.target.value = ""
+    }
+
+    if (variant === "button") {
```

Comment:
For å unngå prop-drilling kan et nyttig pattern være
```tsx
<FileUpload>
  // <FileUpload.DropZone />
  // <FileUpload.UploadButton />
</FileUpload>
```

`FileUpload` vil da inneholde state og deler av logikken + være en context-provider
```tsx
// FileUpload.tsx
...logikk
<FileUploadContext.Provider value={...}>
   {children}
</FileUploadContext.Provider>
```

`FileUpload.DropZone` og `FileUpload.UploadButton` blir da en egne komponenter som har egne props og konsumerer context fra `FileUpload`.